anchor-tx/src/cas_remote.rs

```diff
@@ -0,0 +1,313 @@
+        ))
+    }
+}
+
+impl RemoteCas {
+    /// Create a new RemoteCas instance
+    pub fn new(keypair: Ed25519KeyPair, cas_api_url: String) -> Self {
+        Self {
+            signing_key: keypair,
+            cas_api_url,
+        }
+    }
+
+    /// Create an anchor request on the remote CAS
+    pub async fn create_anchor_request(&self, root_cid: Cid) -> Result<String> {
+        let cas_create_request_url = format!("{}/api/v0/requests", self.cas_api_url);
+        let cas_request_body = serde_json::to_string(&CasAnchorRequest {
+            stream_id: cid_to_stream_id(cid_from_ed25519_key_pair(&self.signing_key)),
+            cid: root_cid.to_string(),
+            timestamp: chrono::Utc::now().to_rfc3339(),
+            ceramic_one_version: AGENT_VERSION.to_owned(),
+        })?;
+        let digest = MultihashDigest::digest(&Code::Sha2_256, cas_request_body.as_bytes());
+        let digest = hex::encode(digest.digest());
+        let auth_jwt = auth_jwt(
+            cas_create_request_url.clone(),
+            format!("0x{}", digest),
+            &self.signing_key,
+        )
+        .await?;
+        let auth_header = format!("Bearer {}", auth_jwt);
+        debug!("auth_header {}", &auth_header);
+        let res = reqwest::Client::new()
+            .post(cas_create_request_url)
+            .header("Authorization", auth_header)
+            .header("Content-Type", "application/json")
+            .body(cas_request_body)
+            .send()
+            .await?;
+        Ok(res.text().await?)
+    }
+}
+
+async fn parse_anchor_response(anchor_response: String) -> Result<CasResponseParseResult> {
+    // Return if we were unable to parse the anchor response
+    let anchor_response = serde_json::from_str::<AnchorResponse>(anchor_response.as_str())?;
+
+    // If the response does not contain a witness CAR file, the anchor request is either still pending or it failed
+    // because the request was unauthorized.
+    let Some(witness_car_b64) = anchor_response.witness_car else {
+        return match anchor_response.message.as_str() {
+            "Unauthorized" => Ok(CasResponseParseResult::Unauthorized),
+            message => {
+                return Err(anyhow!("message from remote CAS: {}", message));
+            }
+        };
+    };
+    let witness_car_bytes = b64_standard.decode(witness_car_b64)?;
+    let car_reader = CarReader::new(witness_car_bytes.as_ref()).await?;
+    let mut remote_merkle_nodes = MerkleNodes::default();
+    let mut detached_time_event: Option<DetachedTimeEvent> = None;
+    let mut proof: Option<Proof> = None;
+    for (cid, block) in car_reader
+        .stream()
+        .into_stream()
+        .try_collect::<Vec<(_, _)>>()
+        .await?
+    {
+        if let Ok(block) = serde_ipld_dagcbor::from_slice::<DetachedTimeEvent>(&block) {
+            detached_time_event = Some(block);
+        }
+        if let Ok(block) = serde_ipld_dagcbor::from_slice::<Proof>(&block) {
```

Comment:
should these last 2 `if let` statements be `else if let`? seems like only 1 will ever pass

Reply:
just noticed these are changed to `else if let` in part 5